tests/helpers/kubernetes/minikube.py

```diff
@@ -1,208 +1,326 @@
-        assert wait_for(p(container_is_running, self.host_client, name), timeout_seconds=timeout, interval_seconds=2), (
-            "timed out waiting for container %s!" % name
-        )
-        self.container = self.host_client.containers.get(name)
-        self.load_kubeconfig(timeout=timeout)
-        self.client = self.get_client()
-        self.get_bootstrapper()
-        self.name = name
-        self.k8s_version = version
-
-    def deploy(self, version, timeout, options=None):
+            ), ("timed out waiting for %s image to be built" % self.image_tag)
+            print("Waited %d seconds" % (time.time() - start_time))
+        print("\nConnecting to cluster in %s container ..." % self.container_name)
+        self.connect_to_cluster(timeout)
+
+    def deploy(self, k8s_version, timeout=300, options=None):
+        self.k8s_version = check_k8s_version(k8s_version)
+        assert self.get_version(k8s_version), "failed to get minikube version"
         if options is None:
             options = {}
-        self.registry_port = get_free_port()
-        if container_is_running(self.host_client, "minikube"):
-            self.host_client.containers.get("minikube").remove(force=True, v=True)
-        self.k8s_version = version
-        if self.k8s_version[0] != "v":
-            self.k8s_version = "v" + self.k8s_version
-        if not options:
-            options = {
-                "name": "minikube",
-                "privileged": True,
-                "environment": {"K8S_VERSION": self.k8s_version, "TIMEOUT": str(timeout)},
-                "ports": {
-                    "8080/tcp": None,
-                    "8443/tcp": None,
-                    "2375/tcp": None,
-                    "%d/tcp" % self.registry_port: self.registry_port,
-                },
-                "volumes": {"/tmp/scratch": {"bind": "/tmp/scratch", "mode": "rw"}},
-            }
-        if MINIKUBE_VERSION:
-            self.version = MINIKUBE_VERSION
-        elif semver.match(self.k8s_version.lstrip("v"), ">=1.11.0"):
-            self.version = MINIKUBE_KUBEADM_VERSION
-        else:
-            self.version = MINIKUBE_LOCALKUBE_VERSION
-        if self.version == "latest" or semver.match(
-            self.version.lstrip("v"), ">" + MINIKUBE_LOCALKUBE_VERSION.lstrip("v")
-        ):
-            options["command"] = "/lib/systemd/systemd"
-        else:
-            options["command"] = "sleep inf"
+        options.setdefault("name", MINIKUBE_CONTAINER_NAME)
+        try:
+            self.host_client.containers.get(options["name"]).remove(force=True, v=True)
+        except docker.errors.NotFound:
+            pass
+        options.setdefault("privileged", True)
+        options.setdefault(
+            "environment",
+            {"K8S_VERSION": self.k8s_version, "TIMEOUT": str(timeout), "KUBECONFIG_PATH": MINIKUBE_KUBECONFIG_PATH},
+        )
+        if tcp_socket_open("127.0.0.1", self.registry_port):
```

Comment:
I don't understand this conditional.  Shouldn't it be the opposite?

Reply:
`tcp_socket_open` returns true if the port is connectable (i.e. in use). So this just assigns a new port for the registry if for some reason the default port is in use.

Reply:
ok, yeah I should have called that function `tcp_socket_bound` in retrospect since "open" is easily confused with "free".